packages/dds/tree/docs/wip/identifiers.md

```diff
@@ -0,0 +1,73 @@
+
+## Updating the Index
+
+Like all indexes, the IdentifierIndex can be serialized and rehydrated. When a SharedTree loads a document, the IdentifierIndex loads its last serialized state. From then on, it is updated as part of every operation:
+
+* Insert: The tree being inserted is fully traversed in a search for identifier fields. Any that are found are added to the index. If other indexes are also traversing the tree for their own purposes, it might make sense to have SharedTree provide a single traversal that the indexes can hook into, rather than having each index do its own walk.
+* Delete: The tree being deleted is fully traversed in a search for identifier fields. Any that are found are removed from the index. For a large, virtualized tree this might require downloading parts of the tree that haven't been loaded yet; therefore, updating the index after a delete becomes an async operation. At first glance, this seems to imply op processing must be async, which is probably not acceptable. There may be alternative strategies; for example, the index could allow the removal of the identifiers to run asynchrounously in the background. Then, for each identifier query, the index validates that the mapping it has is still valid by looking up the node in the forest by its path and then checking that its identifier matches the one in the index.
+* Move: The tree being moved is fully traversed in a search for identifier fields. Any that are found are updated with new paths in the index. This has similar requirements to delete, just above.
+* Set Value: If the value being set is an identifier, it must update the index. This case is not necessary if identifier fields are required to be readonly (and can therefore only be introduced during inserts).
+
+## API
+
+1. An API for querying a node by identifier, e.g. `find(id: CompressedNodeIdentifier): Cursor | undefined`. Higher level reading APIs (like EditableTree) would wrap this for their own purposes, although I don't think there is a way to guarantee the type of the resulting node at compile time, even in a context that is schema aware. This API presents the IdentifierIndex as a covering index; it might be useful to also expose the inner `findPath(id: CompressedNodeIdentifier): UpPath | undefined` so it can be used as a non-covering index as well.
+2. APIs for updating the contents of the index, e.g. `set(id: CompressedNodeIdentifier, path: UpPath)` and `delete(id: CompressedNodeIdentifier)`
```

Comment:
Don't we need a `move(id: CompressedNodeIdentifier, path: UpPath)` as well? I can see how `set` could be used for both, but that we reduce our ability to detect node ID duplication bugs.

Reply:
Would a move would be a:

1. assert id is already present
2. delete the id
3. set the id with new path

And you're saying that step 1 is valuable for debugging? Or is there something else I'm not thinking of?